<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Detect Elevations and Gaps in Mapped Sequencing Read Coverage • ProActive</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" sizes="any" href="favicon.ico">
<link rel="manifest" href="site.webmanifest">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Detect Elevations and Gaps in Mapped Sequencing Read Coverage">
<meta name="description" content="Automate the detection of gaps and elevations in mapped sequencing read coverage using a 2D pattern-matching algorithm. ProActive detects, characterizes and visualizes read coverage patterns in both genomes and metagenomes. Optionally, users may provide gene annotations associated with their genome or metagenome in the form of a .gff file. In this case, ProActive will generate an additional output table containing the gene annotations found within the detected regions of gapped and elevated read coverage. Additionally, users can search for gene annotations of interest in the output read coverage plots.">
<meta property="og:description" content="Automate the detection of gaps and elevations in mapped sequencing read coverage using a 2D pattern-matching algorithm. ProActive detects, characterizes and visualizes read coverage patterns in both genomes and metagenomes. Optionally, users may provide gene annotations associated with their genome or metagenome in the form of a .gff file. In this case, ProActive will generate an additional output table containing the gene annotations found within the detected regions of gapped and elevated read coverage. Additionally, users can search for gene annotations of interest in the output read coverage plots.">
<meta property="og:image" content="https://jlmaier12.github.io/ProActive/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">ProActive</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/ProActiveVignette.html">ProActive</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jlmaier12/ProActive/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="proactive">ProActive<a class="anchor" aria-label="anchor" href="#proactive"></a>
</h1>
</div>
<!-- badges: start -->
<!-- badges: end -->
<p><strong><code>ProActive</code> automatically detects regions of gapped and elevated read coverage using a 2D pattern-matching algorithm. <code>ProActive</code> detects, characterizes and visualizes read coverage patterns in both genomes and metagenomes. Optionally, users may provide gene annotations associated with their genome or metagenome in the form of a .gff file. In this case, <code>ProActive</code> will generate an additional output table containing the gene annotations found within the detected regions of gapped and elevated read coverage. Additionally, users can search for gene annotations of interest in the output read coverage plots.</strong></p>
<p>Visualizing read coverage data is important because gaps and elevations in coverage can be indicators of a variety of biological and non-biological scenarios, for example-</p>
<ul>
<li>Elevations and gaps in read coverage may be caused by some types of structural variants. Deletions can cause gaps while duplications can cause elevations in read coverage [1].</li>
<li>Highly active and/or abundant mobile genetic elements, like transposable elements [2] and prophage [3] for example, can create elevations in read coverage at their respective integration sites.</li>
<li>Genetic regions with high mutation rates and/or high variability within the population can generate gaps in read coverage [4].</li>
<li>Poor quality sequencing reads and chimeric reference sequences may cause gaps and elevations in read coverage.</li>
</ul>
<p><strong>Since the cause for gaps and elevations in read coverage can be ambiguous, ProActive is best used as a screening method to identify genetic regions for further investigation with other tools!</strong></p>
<p><strong>References:</strong></p>
<ol style="list-style-type: decimal">
<li>Tattini L., D’Aurizio R., &amp; Magi A. (2015). Detection of Genomic Structural Variants from Next-Generation Sequencing Data. Frontiers in bioengineering and biotechnology, 3, 92. <a href="https://doi.org/10.3389/fbioe.2015.00092" class="external-link uri">https://doi.org/10.3389/fbioe.2015.00092</a>
</li>
<li>Kleiner M., Bushnell B., Sanderson K.E. et al. (2020) Transductomics: sequencing-based detection and analysis of transduced DNA in pure cultures and microbial communities. Microbiome 8, 158. <a href="https://doi.org/10.1186/s40168-020-00935-5" class="external-link uri">https://doi.org/10.1186/s40168-020-00935-5</a>
</li>
<li>Kieft K., Anantharaman K. (2022). Deciphering Active Prophages from Metagenomes. mSystems 7:e00084-22. <a href="https://doi.org/10.1128/msystems.00084-22" class="external-link uri">https://doi.org/10.1128/msystems.00084-22</a>
</li>
<li>Fogarty E., Moore R. (2019). Visualizing contig coverages to better understand microbial population structure. <a href="https://merenlab.org/2019/11/25/visualizing-coverages/" class="external-link uri">https://merenlab.org/2019/11/25/visualizing-coverages/</a>
</li>
</ol>
<div class="section level3">
<h3 id="input-files">Input files<a class="anchor" aria-label="anchor" href="#input-files"></a>
</h3>
<div class="section level4">
<h4 id="pileup-file">Pileup file:<a class="anchor" aria-label="anchor" href="#pileup-file"></a>
</h4>
<p>ProActive detects read coverage patterns using a pattern-matching algorithm that operates on pileup files. A pileup file is a file format where each row summarizes the ‘pileup’ of reads at specific genomic locations. Pileup files can be used to generate a rolling mean of read coverages and associated base pair positions which reduces data size while preserving read coverage patterns. <strong>ProActive requires that input pileups files</strong> <strong>be generated using a 100 bp window/bin size.</strong></p>
<p>Pileup files can be generated by mapping sequencing reads to a metagenome or genome fasta. <strong>Read mapping should be performed using a high</strong> <strong>minimum identity (0.97 or higher) and random mapping of ambiguous reads.</strong> The pileup files needed for ProActive are generated using the .bam files produced during read mapping. Some read mappers, like <a href="https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbmap-guide/" class="external-link">BBMap</a>, allow for the generation of pileup files in the <a href="https://github.com/BioInfoTools/BBMap/blob/master/sh/bbmap.sh" class="external-link"><code>bbmap.sh</code></a> command with use of the <code>bincov</code> output with the <code>covbinsize=100</code> parameter/argument. <strong>Otherwise, BBMap’s</strong> <strong><a href="https://github.com/BioInfoTools/BBMap/blob/master/sh/pileup.sh" class="external-link"><code>pileup.sh</code></a></strong> <strong>can convert .bam files produced by any read mapper to pileup files</strong> <strong>compatible with ProActive using the <code>bincov</code> output with <code>binsize=100</code>.</strong></p>
<p><strong>NOTE:</strong> For detailed information on input file format, please see the vignette. Users may also use the ‘sampleMetagenomePileup’ and ‘sampleGenomePileup’ files that come pre-loaded with ProActive as a reference.</p>
</div>
<div class="section level4">
<h4 id="gfftsv">gffTSV:<a class="anchor" aria-label="anchor" href="#gfftsv"></a>
</h4>
<p>ProActive optionally accepts a .gff file as input. The .gff file must be associated with the same metagenome or genome used to create your pileup file. The .gff file should be a TSV and should follow the same general format described <a href="https://en.wikipedia.org/wiki/General_feature_format#:~:text=In%20bioinformatics%2C%20the%20general%20feature,DNA%2C%20RNA%20and%20protein%20sequences." class="external-link">here</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Install ProActive from CRAN with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ProActive"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jlmaier12/ProActive" class="external-link">ProActive</a></span><span class="op">)</span></span></code></pre></div>
<p>Install the development version of ProActive from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"jlmaier12/ProActive"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jlmaier12/ProActive" class="external-link">ProActive</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jlmaier12/ProActive" class="external-link">ProActive</a></span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## Metagenome mode</span></span>
<span></span>
<span><span class="va">MetagenomeProActive</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ProActiveDetect.html">ProActiveDetect</a></span><span class="op">(</span></span>
<span>  pileup <span class="op">=</span> <span class="va">sampleMetagenomePileup</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"metagenome"</span>,</span>
<span>  gffTSV <span class="op">=</span> <span class="va">sampleMetagenomegffTSV</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Preparing input file for pattern-matching...</span></span>
<span><span class="co">#&gt; Starting pattern-matching...</span></span>
<span><span class="co">#&gt; A quarter of the way done with pattern-matching</span></span>
<span><span class="co">#&gt; Half of the way done with pattern-matching</span></span>
<span><span class="co">#&gt; Almost done with pattern-matching!</span></span>
<span><span class="co">#&gt; Summarizing pattern-matching results</span></span>
<span><span class="co">#&gt; Finding gene predictions in elevated or gapped regions of read coverage...</span></span>
<span><span class="co">#&gt; Finalizing output</span></span>
<span><span class="co">#&gt; Execution time: 2.09secs</span></span>
<span><span class="co">#&gt; 0 contigs were filtered out based on low read coverage</span></span>
<span><span class="co">#&gt; 0 contigs were filtered out based on length (&lt; minContigLength)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Elevation       Gap NoPattern </span></span>
<span><span class="co">#&gt;         3         3         1</span></span>
<span></span>
<span><span class="va">MetagenomePlots</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plotProActiveResults.html">plotProActiveResults</a></span><span class="op">(</span>pileup <span class="op">=</span> <span class="va">sampleMetagenomePileup</span>,</span>
<span>                                        ProActiveResults <span class="op">=</span> <span class="va">MetagenomeProActive</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MetagenomeGeneMatches</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/geneAnnotationSearch.html">geneAnnotationSearch</a></span><span class="op">(</span>ProActiveResults <span class="op">=</span> <span class="va">MetagenomeProActive</span>, </span>
<span>                                              pileup <span class="op">=</span> <span class="va">sampleMetagenomePileup</span>, </span>
<span>                                              gffTSV <span class="op">=</span> <span class="va">sampleMetagenomegffTSV</span>,</span>
<span>                                              geneOrProduct <span class="op">=</span> <span class="st">"product"</span>,</span>
<span>                                              keyWords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"transport"</span>, <span class="st">"chemotaxis"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Cleaning gff file...</span></span>
<span><span class="co">#&gt; Cleaning pileup file...</span></span>
<span><span class="co">#&gt; Searching for matching annotations...</span></span>
<span><span class="co">#&gt; 3 contigs/chunks have gene annotations that match one or more of the provided keyWords</span></span>
<span></span>
<span></span>
<span><span class="co">## Genome mode</span></span>
<span></span>
<span><span class="va">GenomeProActive</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ProActiveDetect.html">ProActiveDetect</a></span><span class="op">(</span></span>
<span>  pileup <span class="op">=</span> <span class="va">sampleGenomePileup</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  gffTSV <span class="op">=</span> <span class="va">sampleGenomegffTSV</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Preparing input file for pattern-matching...</span></span>
<span><span class="co">#&gt; Starting pattern-matching...</span></span>
<span><span class="co">#&gt; A quarter of the way done with pattern-matching</span></span>
<span><span class="co">#&gt; Half of the way done with pattern-matching</span></span>
<span><span class="co">#&gt; Almost done with pattern-matching!</span></span>
<span><span class="co">#&gt; Summarizing pattern-matching results</span></span>
<span><span class="co">#&gt; Finding gene predictions in elevated or gapped regions of read coverage...</span></span>
<span><span class="co">#&gt; Finalizing output</span></span>
<span><span class="co">#&gt; Execution time: 29.7secs</span></span>
<span><span class="co">#&gt; 0 contigs were filtered out based on low read coverage</span></span>
<span><span class="co">#&gt; 0 contigs were filtered out based on length (&lt; minContigLength)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Elevation       Gap NoPattern </span></span>
<span><span class="co">#&gt;        25         3        21</span></span>
<span></span>
<span><span class="va">GenomePlots</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plotProActiveResults.html">plotProActiveResults</a></span><span class="op">(</span>pileup <span class="op">=</span> <span class="va">sampleGenomePileup</span>,</span>
<span>                                    ProActiveResults <span class="op">=</span> <span class="va">GenomeProActive</span><span class="op">)</span></span>
<span></span>
<span><span class="va">GenomeGeneMatches</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/geneAnnotationSearch.html">geneAnnotationSearch</a></span><span class="op">(</span>ProActiveResults <span class="op">=</span> <span class="va">GenomeProActive</span>, </span>
<span>                                          pileup <span class="op">=</span> <span class="va">sampleGenomePileup</span>, </span>
<span>                                          gffTSV <span class="op">=</span> <span class="va">sampleGenomegffTSV</span>,</span>
<span>                                          geneOrProduct <span class="op">=</span> <span class="st">"product"</span>,</span>
<span>                                          keyWords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ribosomal"</span><span class="op">)</span>, </span>
<span>                                          inGapOrElev <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                          bpRange <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span>
<span><span class="co">#&gt; Cleaning gff file...</span></span>
<span><span class="co">#&gt; Cleaning pileup file...</span></span>
<span><span class="co">#&gt; Searching for matching annotations...</span></span>
<span><span class="co">#&gt; 8 contigs/chunks have gene annotations that match one or more of the provided keyWords</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=ProActive" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/jlmaier12/ProActive/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/jlmaier12/ProActive/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-2" class="external-link">GPL-2</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing ProActive</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Jessie Maier <br><small class="roles"> Author, maintainer, copyright holder </small> <a href="https://orcid.org/0009-0001-8575-5386" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Manuel Kleiner <br><small class="roles"> Author, thesis advisor </small> <a href="https://orcid.org/0000-0001-6904-0287" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jessie Maier, Manuel Kleiner.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
