<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Detect Elevations and Gaps in Mapped Sequencing Read Coverage • ProActive</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Detect Elevations and Gaps in Mapped Sequencing Read Coverage">
<meta name="description" content="The `ProActive` R package automatically detects regions of gapped and elevated read coverage using a pattern-matching algorithm. `ProActive` can detect, characterize and visualize read coverage patterns in both genomes and metagenomes. Optionally, users may provide gene predictions associated with their genome or metagenome in the form of a .gff file. In this case, `ProActive` will generate an additional output table containing the gene predictions found within the detected regions of gapped and elevated read coverage.">
<meta property="og:description" content="The `ProActive` R package automatically detects regions of gapped and elevated read coverage using a pattern-matching algorithm. `ProActive` can detect, characterize and visualize read coverage patterns in both genomes and metagenomes. Optionally, users may provide gene predictions associated with their genome or metagenome in the form of a .gff file. In this case, `ProActive` will generate an additional output table containing the gene predictions found within the detected regions of gapped and elevated read coverage.">
<meta property="og:image" content="https://jlmaier12.github.io/ProActive/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">ProActive</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/ProActiveVignette.html">ProActive</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jlmaier12/ProActive/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="proactive">ProActive<a class="anchor" aria-label="anchor" href="#proactive"></a>
</h1>
</div>
<!-- badges: start -->
<!-- badges: end -->
<p><strong>The <code>ProActive</code> R package automatically detects regions of gapped and elevated read coverage using a pattern-matching algorithm. <code>ProActive</code> can detect, characterize and visualize read coverage patterns in both genomes and metagenomes. Optionally, users may provide gene predictions associated with their genome or metagenome in the form of a .gff file. In this case, <code>ProActive</code> will generate an additional output table containing the gene predictions found within the detected regions of gapped and elevated read coverage. <code>ProActive</code> is best used as a screening method to identify genetic regions for further investigation.</strong></p>
<p>Elevations or gaps in read coverage can be caused by differential abundance of specific genetic elements. For example, an elevation in read coverage may be caused by prophage activation. When a prophage activates and enters the lytic cycle, its genome begins replicating and the ratio of phage:bacterial genomes in the cell begins to increase. Because there are more phage genomes than bacterial genomes, during sequencing more phage reads are generated than bacterial. When these reads are mapped back to their associated reference sequence, the read coverage of the prophage region will be elevated in comparison to the read coverage of the bacterial genome on either side of the prophage. This same principle applies to temperate phage who are highly abundant in the environment as well as other mobile genetic elements that are freely present in the environment at a higher ratio than the originating or ‘host’ genome.</p>
<p>Conversely, a gap in read coverage may indicate genetic heterogeneity in the associated bacterial population. Genetic variants with and without specific genetic elements, like prophage or certain genes, will produce differential abundances of sequencing reads that may form read coverage gaps. The formation of read coverage gaps due to genetic variants is dependent on the assembly (i.e. if the assembler assembles the genetic variants as separate entities or not). Gaps in read coverage may also form at regions with high mutation rates.</p>
<div class="section level3">
<h3 id="input-files">Input files<a class="anchor" aria-label="anchor" href="#input-files"></a>
</h3>
<div class="section level4">
<h4 id="pileup-file">Pileup file:<a class="anchor" aria-label="anchor" href="#pileup-file"></a>
</h4>
<p>ProActive detects read coverage patterns using a pattern-matching algorithm that operates on pileup files. A pileup file is a file format where each row summarizes the ‘pileup’ of reads at specific genomic locations. Pileup files can be used to generate a rolling mean of read coverages and associated base pair positions which reduces data size while preserving read coverage patterns. <strong>ProActive requires that input pileups files</strong> <strong>be generated using a 100 bp window/bin size.</strong></p>
<p>Pileup files can be generated by mapping sequencing reads to a metagenome or genome fasta. <strong>Read mapping should be performed using a high</strong> <strong>minimum identity (0.97 or higher) and random mapping of ambiguous reads.</strong> The pileup files needed for ProActive are generated using the .bam files produced during read mapping. Some read mappers, like <a href="https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbmap-guide/" class="external-link">BBMap</a>, allow for the generation of pileup files in the <a href="https://github.com/BioInfoTools/BBMap/blob/master/sh/bbmap.sh" class="external-link"><code>bbmap.sh</code></a> command with use of the <code>bincov</code> output with the <code>covbinsize=100</code> parameter/argument. <strong>Otherwise, BBMap’s</strong> <strong><a href="https://github.com/BioInfoTools/BBMap/blob/master/sh/pileup.sh" class="external-link"><code>pileup.sh</code></a></strong> <strong>can convert .bam files produced by any read mapper to pileup files</strong> <strong>compatible with ProActive using the <code>bincov</code> output with <code>binsize=100</code>.</strong></p>
<p><strong>NOTE:</strong> For detailed information on input file format, please see the vignette. Users may also use the ‘sampleMetagenomePileup’ and ‘sampleGenomePileup’ files that come pre-loaded with ProActive as a reference.</p>
</div>
<div class="section level4">
<h4 id="gfftsv">gffTSV:<a class="anchor" aria-label="anchor" href="#gfftsv"></a>
</h4>
<p>ProActive optionally accepts a .gff file as input. The .gff file must be associated with the same metagenome or genome used to create your pileup file. The .gff file should be a TSV and should follow the same general format described <a href="https://en.wikipedia.org/wiki/General_feature_format#:~:text=In%20bioinformatics%2C%20the%20general%20feature,DNA%2C%20RNA%20and%20protein%20sequences." class="external-link">here</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Install ProActive from CRAN with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ProActive"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jlmaier12/ProActive" class="external-link">ProActive</a></span><span class="op">)</span></span></code></pre></div>
<p>Install the development version of ProActive from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"jlmaier12/ProActive"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jlmaier12/ProActive" class="external-link">ProActive</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>Metagenome mode:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jlmaier12/ProActive" class="external-link">ProActive</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">MetagenomeProActive</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ProActive.html">ProActive</a></span><span class="op">(</span></span>
<span>  pileup <span class="op">=</span> <span class="va">sampleMetagenomePileup</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"metagenome"</span>,</span>
<span>  gffTSV <span class="op">=</span> <span class="va">sampleMetagenomegffTSV</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Preparing input file for pattern-matching...</span></span>
<span><span class="co">#&gt; Starting pattern-matching...</span></span>
<span><span class="co">#&gt; A quarter of the way done with pattern-matching</span></span>
<span><span class="co">#&gt; Half of the way done with pattern-matching</span></span>
<span><span class="co">#&gt; Almost done with pattern-matching!</span></span>
<span><span class="co">#&gt; Summarizing pattern-matching results</span></span>
<span><span class="co">#&gt; Finding gene predictions in elevated or gapped regions of read coverage...</span></span>
<span><span class="co">#&gt; Finalizing output</span></span>
<span><span class="co">#&gt; Execution time: 1.98secs</span></span>
<span><span class="co">#&gt; 0 contigs were filtered out based on low read coverage</span></span>
<span><span class="co">#&gt; 0 contigs were filtered out based on length (&lt; minContigLength)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Elevation       Gap NoPattern </span></span>
<span><span class="co">#&gt;         3         3         1</span></span>
<span></span>
<span><span class="va">MetagenomePlots</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plotProActiveResults.html">plotProActiveResults</a></span><span class="op">(</span>pileup <span class="op">=</span> <span class="va">sampleMetagenomePileup</span>,</span>
<span>                                        ProActiveResults <span class="op">=</span> <span class="va">MetagenomeProActive</span><span class="op">)</span></span></code></pre></div>
<p>Genome mode:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GenomeProActive</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ProActive.html">ProActive</a></span><span class="op">(</span></span>
<span>  pileup <span class="op">=</span> <span class="va">sampleGenomePileup</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  gffTSV <span class="op">=</span> <span class="va">sampleGenomegffTSV</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Preparing input file for pattern-matching...</span></span>
<span><span class="co">#&gt; Starting pattern-matching...</span></span>
<span><span class="co">#&gt; A quarter of the way done with pattern-matching</span></span>
<span><span class="co">#&gt; Half of the way done with pattern-matching</span></span>
<span><span class="co">#&gt; Almost done with pattern-matching!</span></span>
<span><span class="co">#&gt; Summarizing pattern-matching results</span></span>
<span><span class="co">#&gt; Finding gene predictions in elevated or gapped regions of read coverage...</span></span>
<span><span class="co">#&gt; Finalizing output</span></span>
<span><span class="co">#&gt; Execution time: 36.98secs</span></span>
<span><span class="co">#&gt; 0 contigs were filtered out based on low read coverage</span></span>
<span><span class="co">#&gt; 0 contigs were filtered out based on length (&lt; minContigLength)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Elevation       Gap NoPattern </span></span>
<span><span class="co">#&gt;        25         3        21</span></span>
<span></span>
<span><span class="va">GenomePlots</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plotProActiveResults.html">plotProActiveResults</a></span><span class="op">(</span>pileup <span class="op">=</span> <span class="va">sampleGenomePileup</span>,</span>
<span>                                    ProActiveResults <span class="op">=</span> <span class="va">GenomeProActive</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/jlmaier12/ProActive/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/jlmaier12/ProActive/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-2" class="external-link">GPL-2</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing ProActive</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Jessie Maier <br><small class="roles"> Author, maintainer, copyright holder </small> <a href="https://orcid.org/0009-0001-8575-5386" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Manuel Kleiner <br><small class="roles"> Author, thesis advisor </small> <a href="https://orcid.org/0000-0001-6904-0287" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jessie Maier, Manuel Kleiner.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
