![ProActive](https://github.com/jlmaier12/ProActive/assets/45083046/4d3a0ab9-1232-4b6f-935f-1e84691b452d)

Detect elevated read coverages on contigs that may be associated with active/highly abundant prophages or other mobile genetic elements (MGEs), like transposases and translocases. 

### Decription
`ProActive` detects regions of elevated read coverage on contigs. When a prophage activates and enters the lytic cycle, its genome begins replicating and the ratio of phage:bacterial genomes in the cell begins to increase. Because there are more phage genomes than bacterial genomes, during sequencing more phage reads are generated than bacterial. When these reads are mapped back to their associated contig, the read coverage of the prophage region will be elevated in comparison to the read coverage of the bacterial genome on either side of the prophage. This same principle applies to temperate phage who are highly abundant in the environment as well as other mobile genetic elements that are freely present in the environment at a higher ratio than the originiating or 'host' genome. 

Figure 1 from Kieft and Anantharaman (2022) is an excellent visualization of this phenomenon with active prophage:

![msystems 00084-22-f001](https://github.com/jlmaier12/ProActive/assets/45083046/7f1d4e54-8ae8-406e-940e-5da311718dba)

**Reference** Kieft K, Anantharaman K. Deciphering Active Prophages from Metagenomes. mSystems. 2022 Apr 26;7(2):e0008422. doi: 10.1128/msystems.00084-22. Epub 2022 Mar 24. PMID: 35323045; PMCID: PMC9040807.

The increase in read coverage at prophage or MGE locations can be assessed and inferences can be made about associated activity/abundance, however, in order to assess elevation of read coverage, the genomic coordinates of the prophage or MGE must be known. Locations of these genetic elements are usually idenfitied, at least in part, using gene annotations which means unannotated prophage or MGEs are missed. ProActive, however, is reference-independent and therefore bypasses this barrier to active prophage identification.

ProActive works by detecting the 'block-like' pattern generated by regions of elevated read coverages. ProActive will scan the mapped read coverages of a genome or metagenome contigs and search for a 'block-like' elevated read coverage pattern. For your genome or metagenome, ProActive generates a summary table containing the sizes of the elevated read coverage regions, the associated start and stop positions, and the associated ratio of elevated:non-elevated read coverage. 

ProActive's read coverage pattern-matching is only as good as the provided data. There are other situations that can create read coverage patterns similar to that of active/abundant prophage or MGEs. For example, chimeric assemblies or contaminants can also create regions of elevated read coverage. Manual curation of ProActive's results is required, however, ProActive has several metrics with which a user can filter results to ease the manual curation process. In addition, ProActive's pattern-matching search can be modified with several user-defined input parameters. See the tutorial below for details. 

NOTE: It is a good idea to use ProActive in combination with other prophage or MGE finding tools. ProActive will not identify dormant or low-abundance prophages or MGEs as the associated reads will likely not not be abundant enought to create a significant elevation in read coverage. 


### Data input for `ProActive`

Map your sequencing read to the associated genome or metagenome contigs. Any read mapper can be used, however we recommend using [BBMap](https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/installation-guide/) with `ambiguous=random`, `qtrim=lr`, and `minid=0.97`. 

Create pileup files using `BBMap` `pileup.sh` with 100 bp window sizes:
```{bash}
$ pileup.sh in=YOUR_SORTED_INDEXED_READ_MAPPING.bam out=output.pileupcovstats bincov=pileup_bincov100.txt binsize=100 stdev=t
```
The pileup_bincov100.txt file produced with `pileup.sh` can be used directly as input for ProActive.

NOTE: ProActive will filter out contigs shorter than 30kbp. If your metagenome assembly consists of majority short contigs, ProActive may not be the right tool to use. 

### Install and run `ProActive`

##### Install `ProActive`:

```R
library(devtools)
install_github("jlmaier12/ProActive", ref="master")
library(ProActive)
```

##### Run `ProActive`:

Import your pileup file:
```R
read_coverages <- read.delim("pileup_bincov100.txt",  header=FALSE, comment.char="#")
```

Run `ProActive()`:
```R
ProActiveResults <- ProActive(pileup=cage1readmapping, mode="metagenome", windowsize=2000, minsize=10000, maxsize=80000, chunksize=NA, cleanup=TRUE) 
```
Input parameters:
- pileup: Your pileup file with 100bp binsizes (aka windowsize)
- mode: Either "genome" or "metagenome". 
- windowsize: Either 200, 500, 1000 and 2000. The number of base pairs to average coverage values over. Larger window sizes improve processing time but the resolution of read coverage patterns may be lost. 1000bp windowsize is the default.
- minsize: The minimum size (in base pairs) of elevated read coverage you would like ProActive to search for. Default is 10000.
- maxsize: The maximum size (in base pairs) of elevated read coverage you would like ProActive to search for. Default is NA (i.e no upper limit set).
- chunksize: If mode="genome", ProActive will 'chunk' the genome to search for a region of elevated read coverage on each chunk. ProActive can only identify one elevated region per chunk!  
- cleanup: Either TRUE or FALSE. ProActive will clean and reformat your input pileup_bincov100.txt file for proper compatibility with functions of ProActive if cleanup=TRUE. Default is TRUE. 

The output of `ProActive()` is a list. Assign list items of interest to their own variables for further use:
```R
summarytable <- ProActiveResults$SummaryTable
VC_PatternMatches <- ProActiveResults$VeryConfidentPatternMatches
C_PatternMatches <- ProActiveResults$ConfidentPatternMatches
NC_PatternMatches <- ProActiveResults$NotConfidentPatternMatches
FilterOutSummaryTable <- ProActiveResults$FilteredOut
```
- ProActiveResults$SummaryTable
- ProActiveResults$VeryConfidentPatternMatches
- ProActiveResults$ConfidentPatternMatches
- ProActiveResults$NotConfidentPatternMatches
- ProActiveResults$FilteredOut: A dataframe containing the contigs or genome chunks that were filtered out due to being shorter than 30kbp or having low read coverage. 

Plot mapped read coverage plots of contigs or genome chunks detected with regions of elevated read coverage. The 'block-like' pattern that ProActive used to make its prediction for each contig/genome chunk will be overlayed on the barplot in black:
```R
#plot confident predictions
ProActivePredictionPlots(pileup=cage1readmapping, ProActiveResults=ProActiveResults, ProActive_shapelist=C_PatternMatches, elevation_filter=1.5, cleanup=TRUE)
```
- pileup
- ProActiveResults
- ProActive_shapelist
- elevation_filter
- cleanup

Search for gene annotations of interest:

gff file format:
Depending on how your gff file was generated, it may contain additional information above and below the table containing gene annotations. To import your gff file to R, it must have this additional inforamtion removed. The easiest way to do this is to select only the necessary information from your gff and copy it to a new file. This can be done on the command line using the following code:
```bash
$ grep 
```

```R
GeneAnnotationSearch <- function(ProActiveResults, ProActive_shapelist, pileup, gene_annots, geneorproduct, keywords, genelocation="nonspecific", bprange = 0, cleanup=TRUE, specificcontig) 
```
