<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Detect elevations and gaps in mapped read coverage patterns. — ProActive • ProActive</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Detect elevations and gaps in mapped read coverage patterns. — ProActive"><meta name="description" content="Performs read coverage pattern-matching and summarizes the results into a list.
 The first list item summarizes the pattern-matching results. The second list
item is the 'cleaned' version of the summary table with all the 'noPattern'
classifications removed. (i.e were not filtered out). The third list item contains
the pattern-match information needed for pattern-match visualization with
`plotProActiveResults()`. The fourth list item is a table containing all the contigs that
were filtered out prior to pattern-matching. The fifth list item contains arguments used during
pattern-matching (windowSize, mode, chunkSize, chunkContigs). If the user provides a
gffTSV files, then the last list is a table consisting of ORFs found within
the detected gaps and elevations in read coverage."><meta property="og:description" content="Performs read coverage pattern-matching and summarizes the results into a list.
 The first list item summarizes the pattern-matching results. The second list
item is the 'cleaned' version of the summary table with all the 'noPattern'
classifications removed. (i.e were not filtered out). The third list item contains
the pattern-match information needed for pattern-match visualization with
`plotProActiveResults()`. The fourth list item is a table containing all the contigs that
were filtered out prior to pattern-matching. The fifth list item contains arguments used during
pattern-matching (windowSize, mode, chunkSize, chunkContigs). If the user provides a
gffTSV files, then the last list is a table consisting of ORFs found within
the detected gaps and elevations in read coverage."><meta property="og:image" content="https://jlmaier12.github.io/ProActive/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ProActive</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/ProActiveVignette.html">ProActive</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jlmaier12/ProActive/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Detect elevations and gaps in mapped read coverage patterns.</h1>
      <small class="dont-index">Source: <a href="https://github.com/jlmaier12/ProActive/blob/main/R/ProActive.R" class="external-link"><code>R/ProActive.R</code></a></small>
      <div class="d-none name"><code>ProActive.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Performs read coverage pattern-matching and summarizes the results into a list.
 The first list item summarizes the pattern-matching results. The second list
item is the 'cleaned' version of the summary table with all the 'noPattern'
classifications removed. (i.e were not filtered out). The third list item contains
the pattern-match information needed for pattern-match visualization with
`plotProActiveResults()`. The fourth list item is a table containing all the contigs that
were filtered out prior to pattern-matching. The fifth list item contains arguments used during
pattern-matching (windowSize, mode, chunkSize, chunkContigs). If the user provides a
gffTSV files, then the last list is a table consisting of ORFs found within
the detected gaps and elevations in read coverage.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ProActive</span><span class="op">(</span></span>
<span>  <span class="va">pileup</span>,</span>
<span>  <span class="va">mode</span>,</span>
<span>  <span class="va">gffTSV</span>,</span>
<span>  windowSize <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  chunkContigs <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  minSize <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  maxSize <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  minContigLength <span class="op">=</span> <span class="fl">30000</span>,</span>
<span>  chunkSize <span class="op">=</span> <span class="fl">1e+05</span>,</span>
<span>  IncludeNoPatterns <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">saveFilesTo</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-pileup">pileup<a class="anchor" aria-label="anchor" href="#arg-pileup"></a></dt>
<dd><p>A .txt file containing mapped sequencing read coverages averaged over
100 bp windows/bins.</p></dd>


<dt id="arg-mode">mode<a class="anchor" aria-label="anchor" href="#arg-mode"></a></dt>
<dd><p>Either "genome" or "metagenome"</p></dd>


<dt id="arg-gfftsv">gffTSV<a class="anchor" aria-label="anchor" href="#arg-gfftsv"></a></dt>
<dd><p>Optional, a .gff file (TSV) containing gene predictions associated with the .fasta
file used to generate the pileup.</p></dd>


<dt id="arg-windowsize">windowSize<a class="anchor" aria-label="anchor" href="#arg-windowsize"></a></dt>
<dd><p>The number of basepairs to average read coverage values over.
Options are 100, 200, 500, 1000 ONLY. Default is 1000.</p></dd>


<dt id="arg-chunkcontigs">chunkContigs<a class="anchor" aria-label="anchor" href="#arg-chunkcontigs"></a></dt>
<dd><p>TRUE or FALSE, If TRUE and `mode`="metagenome", contigs longer
than the `chunkSize` will be 'chunked' into smaller subsets and pattern-matching
will be performed on each subset. Default is FALSE.</p></dd>


<dt id="arg-minsize">minSize<a class="anchor" aria-label="anchor" href="#arg-minsize"></a></dt>
<dd><p>The minimum size (in bp) of elevation or gap patterns. Default is 10000.</p></dd>


<dt id="arg-maxsize">maxSize<a class="anchor" aria-label="anchor" href="#arg-maxsize"></a></dt>
<dd><p>The maximum size (in bp) of elevation or gap patterns. Default is NA
(i.e. no maximum).</p></dd>


<dt id="arg-mincontiglength">minContigLength<a class="anchor" aria-label="anchor" href="#arg-mincontiglength"></a></dt>
<dd><p>The minimum contig/chunk size (in bp) to perform pattern-matching
on. Default is 25000.</p></dd>


<dt id="arg-chunksize">chunkSize<a class="anchor" aria-label="anchor" href="#arg-chunksize"></a></dt>
<dd><p>If `mode`="genome" OR if `mode`="metagenome" and `chunkContigs`=TRUE,
chunk the genome or contigs, respectively, into smaller subsets for pattern-matching.
`chunkSize` determines the size (in bp) of each 'chunk'. Default is 100000.</p></dd>


<dt id="arg-includenopatterns">IncludeNoPatterns<a class="anchor" aria-label="anchor" href="#arg-includenopatterns"></a></dt>
<dd><p>TRUE or FALSE, If TRUE the noPattern pattern-matches will
be included in the ProActive PatternMatches output list. If you would like to visualize
the noPattern pattern-matches in `plotProActiveResults()`, this should be set to TRUE.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>TRUE or FALSE. Print progress messages to console. Default is TRUE.</p></dd>


<dt id="arg-savefilesto">saveFilesTo<a class="anchor" aria-label="anchor" href="#arg-savefilesto"></a></dt>
<dd><p>Optional, Provide a path to the directory you wish to save
output to. A folder will be made within the provided directory to store
results.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing 6 objects described in the function description.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">metagenome_results</span> <span class="op">&lt;-</span> <span class="fu">ProActive</span><span class="op">(</span></span></span>
<span class="r-in"><span>  pileup <span class="op">=</span> <span class="va">sampleMetagenomePileup</span>,</span></span>
<span class="r-in"><span>  mode <span class="op">=</span> <span class="st">"metagenome"</span>,</span></span>
<span class="r-in"><span>  gffTSV <span class="op">=</span> <span class="va">sampleMetagenomegffTSV</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Preparing input file for pattern-matching...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Starting pattern-matching...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> A quarter of the way done with pattern-matching</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Half of the way done with pattern-matching</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Almost done with pattern-matching!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Summarizing pattern-matching results</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finding gene predictions in elevated or gapped regions of read coverage...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finalizing output</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Execution time: 2.34secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0 contigs were filtered out based on low read coverage</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0 contigs were filtered out based on length (&lt; minContigLength)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Elevation       Gap NoPattern </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>         3         3         1 </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jessie Maier, Manuel Kleiner.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

