<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>ProActive • ProActive</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="ProActive">
<meta name="description" content="The `ProActive` R package automatically detects regions of gapped and elevated  read coverage using a pattern-matching algorithm. `ProActive` can detect, characterize  and visualize read coverage patterns in both genomes and metagenomes. Optionally,  users may provide gene annotations associated with their genome or metagenome  in the form of a .gff file. In this case, `ProActive` will generate an additional  output table containing the annotations found within the detected regions of gapped  and elevated read coverage.
">
<meta property="og:description" content="The `ProActive` R package automatically detects regions of gapped and elevated  read coverage using a pattern-matching algorithm. `ProActive` can detect, characterize  and visualize read coverage patterns in both genomes and metagenomes. Optionally,  users may provide gene annotations associated with their genome or metagenome  in the form of a .gff file. In this case, `ProActive` will generate an additional  output table containing the annotations found within the detected regions of gapped  and elevated read coverage.
">
<meta property="og:image" content="https://jlmaier12.github.io/ProActive/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ProActive</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/ProActiveVignette.html">ProActive</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jlmaier12/ProActive/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<script src="ProActiveVignette_files/kePrint-0.0.1/kePrint.js"></script><link href="ProActiveVignette_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>ProActive</h1>
                        <h4 data-toc-skip class="author">Jessie
Maier</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jlmaier12/ProActive/blob/main/vignettes/ProActiveVignette.Rmd" class="external-link"><code>vignettes/ProActiveVignette.Rmd</code></a></small>
      <div class="d-none name"><code>ProActiveVignette.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><strong><code>ProActive</code> automatically detects regions of
gapped and elevated read coverage using a 2D pattern-matching algorithm.
<code>ProActive</code> detects, characterizes and visualizes read
coverage patterns in both genomes and metagenomes. Optionally, users may
provide gene annotations associated with their genome or metagenome in
the form of a .gff file. In this case, <code>ProActive</code> will
generate an additional output table containing the gene annotations
found within the detected regions of gapped and elevated read coverage.
Additionally, users can search for gene annotations of interest in the
output read coverage plots.</strong></p>
<p>Visualizing read coverage data is important because gaps and
elevations in coverage can be indicators of a variety of biological and
non-biological scenarios, for example-</p>
<ul>
<li>Elevations and gaps in read coverage may be caused by some types of
structural variants. Deletions can cause gaps while duplications can
cause elevations in read coverage [1].</li>
<li>Highly active and/or abundant mobile genetic elements, like
transposable elements [2] and prophage [3] for example, can create
elevations in read coverage at their respective integration sites.</li>
<li>Genetic regions with high mutation rates and/or high variability
within the population can generate gaps in read coverage [4].</li>
<li>Poor quality sequencing reads and chimeric reference sequences may
cause gaps and elevations in read coverage.</li>
</ul>
<p><strong>Since the cause for gaps and elevations in read coverage can
be ambiguous, ProActive is best used as a screening method to identify
genetic regions for further investigation with other tools!</strong></p>
<p><strong>References:</strong></p>
<ol style="list-style-type: decimal">
<li>Tattini L., D’Aurizio R., &amp; Magi A. (2015). Detection of Genomic
Structural Variants from Next-Generation Sequencing Data. Frontiers in
bioengineering and biotechnology, 3, 92. <a href="https://doi.org/10.3389/fbioe.2015.00092" class="external-link uri">https://doi.org/10.3389/fbioe.2015.00092</a>
</li>
<li>Kleiner M., Bushnell B., Sanderson K.E. et al. (2020)
Transductomics: sequencing-based detection and analysis of transduced
DNA in pure cultures and microbial communities. Microbiome 8, 158. <a href="https://doi.org/10.1186/s40168-020-00935-5" class="external-link uri">https://doi.org/10.1186/s40168-020-00935-5</a>
</li>
<li>Kieft K., Anantharaman K. (2022). Deciphering Active Prophages from
Metagenomes. mSystems 7:e00084-22.
https://doi.org/10.1128/msystems.00084-22</li>
<li>Fogarty E., Moore R. (2019). Visualizing contig coverages to better
understand microbial population structure. <a href="https://merenlab.org/2019/11/25/visualizing-coverages/" class="external-link uri">https://merenlab.org/2019/11/25/visualizing-coverages/</a>
</li>
</ol>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="cran-install">CRAN install<a class="anchor" aria-label="anchor" href="#cran-install"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ProActive"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jlmaier12/ProActive" class="external-link">ProActive</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="github-install">GitHub install<a class="anchor" aria-label="anchor" href="#github-install"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"jlmaier12/ProActive"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jlmaier12/ProActive" class="external-link">ProActive</a></span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h2>
<div class="section level3">
<h3 id="pileups">Pileups<a class="anchor" aria-label="anchor" href="#pileups"></a>
</h3>
<p>ProActive detects read coverage patterns using a pattern-matching
algorithm that operates on pileup files. A pileup file is a file format
where each row summarizes the ‘pileup’ of reads at specific genomic
locations. Pileup files can be used to generate a rolling mean of read
coverages and associated base pair positions which reduces data size
while preserving read coverage patterns. <strong>ProActive requires that
input pileups files</strong> <strong>be generated using a 100 bp
window/bin size.</strong></p>
<p>Pileup files are generated using the .bam files produced after
mapping sequencing reads to a metagenome or genome fasta file.
<strong>Read mapping should be performed using a high</strong>
<strong>minimum identity (0.97 or higher) and random mapping of
ambiguous reads.</strong></p>
<p>Some read mappers, like <a href="https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbmap-guide/" class="external-link">BBMap</a>,
allow for the generation of pileup files in the <a href="https://github.com/BioInfoTools/BBMap/blob/master/sh/bbmap.sh" class="external-link"><code>bbmap.sh</code></a>
command with use of the <code>bincov</code> output with the
<code>covbinsize=100</code> parameter/argument. <strong>Otherwise,
BBMap’s</strong> <strong><a href="https://github.com/BioInfoTools/BBMap/blob/master/sh/pileup.sh" class="external-link"><code>pileup.sh</code></a></strong>
<strong>can convert .bam files produced by any read mapper to pileup
files compatible</strong> <strong>with ProActive using the
<code>bincov</code> output with <code>binsize=100</code>.</strong></p>
<p><strong>The input pileup file for metagenomes must have the following
format:</strong></p>
<p>Dataframe with four columns:</p>
<ul>
<li>V1: Contig accession</li>
<li>V2: Mapped read coverage values averaged over 100 bp windows</li>
<li>V3: Starting position (bp) of each 100 bp window. Restarts from 100
at the start of each new contig.</li>
<li>V4: Starting position (bp) of each 100 bp window. Does NOT restart
at the start of each new contig.</li>
</ul>
<table class="table table" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
V1
</th>
<th style="text-align:right;">
V2
</th>
<th style="text-align:right;">
V3
</th>
<th style="text-align:right;">
V4
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
NODE_1911 length_44214_cov_4.82142_ID_9560073
</td>
<td style="text-align:right;">
54.66
</td>
<td style="text-align:right;">
100
</td>
<td style="text-align:right;">
175075473
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1911 length_44214_cov_4.82142_ID_9560073
</td>
<td style="text-align:right;">
59.13
</td>
<td style="text-align:right;">
200
</td>
<td style="text-align:right;">
175075573
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1911 length_44214_cov_4.82142_ID_9560073
</td>
<td style="text-align:right;">
53.99
</td>
<td style="text-align:right;">
300
</td>
<td style="text-align:right;">
175075673
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1911 length_44214_cov_4.82142_ID_9560073
</td>
<td style="text-align:right;">
54.69
</td>
<td style="text-align:right;">
400
</td>
<td style="text-align:right;">
175075773
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1911 length_44214_cov_4.82142_ID_9560073
</td>
<td style="text-align:right;">
54.40
</td>
<td style="text-align:right;">
500
</td>
<td style="text-align:right;">
175075873
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1911 length_44214_cov_4.82142_ID_9560073
</td>
<td style="text-align:right;">
52.11
</td>
<td style="text-align:right;">
600
</td>
<td style="text-align:right;">
175075973
</td>
</tr>
</tbody>
</table>
<p><strong>Note that the format for a genome pileup will be slightly
different! The third</strong> <strong>column (V3) does not restart and
the fourth column (V4) starts from 0. ProActive</strong>
<strong>accounts for the differences in pileup formats between genomes
and metagenomes.</strong></p>
<p>Users may use the ‘sampleMetagenomePileup’ and ‘sampleGenomePileup’
files that come pre-loaded with ProActive as references for proper input
file format.</p>
</div>
<div class="section level3">
<h3 id="gff-tsv">gff TSV<a class="anchor" aria-label="anchor" href="#gff-tsv"></a>
</h3>
<p>Optionally, ProActive will accept a .gff file as additional input.
The .gff file must be associated with the same metagenome or genome used
to create your pileup file. The .gff file should be in TSV format and
should follow the same layout described <a href="https://en.wikipedia.org/wiki/General_feature_format#:~:text=In%20bioinformatics%2C%20the%20general%20feature,DNA%2C%20RNA%20and%20protein%20sequences." class="external-link">here</a>.</p>
<strong>The input .gff file must have the following format
exactly:</strong>
<table class="table table" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
V1
</th>
<th style="text-align:left;">
V2
</th>
<th style="text-align:left;">
V3
</th>
<th style="text-align:right;">
V4
</th>
<th style="text-align:right;">
V5
</th>
<th style="text-align:left;">
V6
</th>
<th style="text-align:left;">
V7
</th>
<th style="text-align:left;">
V8
</th>
<th style="text-align:left;">
V9
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
NODE_1911
</td>
<td style="text-align:left;">
Prodigal:002006
</td>
<td style="text-align:left;">
CDS
</td>
<td style="text-align:right;">
318
</td>
<td style="text-align:right;">
1166
</td>
<td style="text-align:left;">
.
</td>
<td style="text-align:left;">
<ul><li></ul>
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
ID=NJKKNKEE_164175;inference=ab initio
prediction:Prodigal:002006;locus_tag=NJKKNKEE_164175;product=hypothetical
protein
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1911
</td>
<td style="text-align:left;">
Prodigal:002006
</td>
<td style="text-align:left;">
CDS
</td>
<td style="text-align:right;">
1198
</td>
<td style="text-align:right;">
1938
</td>
<td style="text-align:left;">
.
</td>
<td style="text-align:left;">
<ul><li></ul>
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
ID=NJKKNKEE_164176;inference=ab initio
prediction:Prodigal:002006;locus_tag=NJKKNKEE_164176;product=hypothetical
protein
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1911
</td>
<td style="text-align:left;">
Prodigal:002006
</td>
<td style="text-align:left;">
CDS
</td>
<td style="text-align:right;">
1938
</td>
<td style="text-align:right;">
2582
</td>
<td style="text-align:left;">
.
</td>
<td style="text-align:left;">
<ul><li></ul>
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
ID=NJKKNKEE_164177;inference=ab initio
prediction:Prodigal:002006;locus_tag=NJKKNKEE_164177;product=hypothetical
protein
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1911
</td>
<td style="text-align:left;">
Prodigal:002006
</td>
<td style="text-align:left;">
CDS
</td>
<td style="text-align:right;">
2722
</td>
<td style="text-align:right;">
3561
</td>
<td style="text-align:left;">
.
</td>
<td style="text-align:left;">
<ul><li></ul>
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
ID=NJKKNKEE_164178;inference=ab initio
prediction:Prodigal:002006;locus_tag=NJKKNKEE_164178;product=hypothetical
protein
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1911
</td>
<td style="text-align:left;">
Prodigal:002006
</td>
<td style="text-align:left;">
CDS
</td>
<td style="text-align:right;">
3671
</td>
<td style="text-align:right;">
4063
</td>
<td style="text-align:left;">
.
</td>
<td style="text-align:left;">
<ul><li></ul>
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
ID=NJKKNKEE_164179;inference=ab initio
prediction:Prodigal:002006;locus_tag=NJKKNKEE_164179;product=hypothetical
protein
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1911
</td>
<td style="text-align:left;">
Prodigal:002006
</td>
<td style="text-align:left;">
CDS
</td>
<td style="text-align:right;">
4128
</td>
<td style="text-align:right;">
4670
</td>
<td style="text-align:left;">
.
</td>
<td style="text-align:left;">
<ul><li></ul>
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
ID=NJKKNKEE_164180;eC_number=1.11.1.1;Name=rbr1_24;db_xref=COG:COG1592;gene=rbr1_24;inference=ab
initio prediction:Prodigal:002006,similar to AA
sequence:UniProtKB:Q97FZ9;locus_tag=NJKKNKEE_164180;product=Rubrerythrin-1
</td>
</tr>
</tbody>
</table>
<p>(<strong>Hint</strong>- if you are using a gff file output by <a href="https://github.com/tseemann/prokka" class="external-link">PROKKA</a>, you may need to
remove some unnecessary (for ProActive) lines of text at the top of the
file. There are various ways one can remove these additional lines,
however, a nice command-line solution is:)</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">grep</span> ^COMMONID metagenomeAnnots.gff <span class="op">&gt;</span> metagenomeAnnotsForProActive.gff</span></code></pre></div>
<p>The ‘COMMONID’ should be a value that all of your contig or genome
accessions start with. For example, the ‘COMMONID’ for the contig
accessions in the sampleMetagenomegffTSV displayed above could be “NODE”
since all the accessions start with “NODE”.</p>
</div>
</div>
<div class="section level2">
<h2 id="proactivedetect">ProActiveDetect()<a class="anchor" aria-label="anchor" href="#proactivedetect"></a>
</h2>
<p><code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code> is the main function in the ProActive
R package. This function filters contigs/chunks based on length and read
coverage, performs pattern-matching to detect gaps and elevations in
read coverage, identifies start and stop positions and sizes of
pattern-matches, and, optionally, extracts gene annotations that fall
within detect gaps and elevations in read coverage.</p>
<div class="section level3">
<h3 id="function-components">Function components<a class="anchor" aria-label="anchor" href="#function-components"></a>
</h3>
<div class="section level4">
<h4 id="chunking">Chunking<a class="anchor" aria-label="anchor" href="#chunking"></a>
</h4>
<p>Currently, <code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code> can only detect one gap or
elevation pattern per contig. Until <code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code> is
able to detect multiple read coverage patterns per contig, we
implemented a ‘chunking’ functionality which (if
<code>chunkContigs</code> = TRUE) chunks large contigs into smaller
subsets (defined by <code>chunkSize</code>) so that pattern-matching can
be performed on each chunk as if it were an individual contig. The
chunking mechanism is what allows <code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code> to
perform pattern-matching on entire genome sequences. When
contigs/genomes are chunked, they are assigned a sequential value to
link chunks back together (i.e. “NODE_1_chunk_1, NODE_1_chunk_2,
NODE_1_chunk_3, …). Note that the remaining ‘chunk’ of a contig/genome
may not be long enough to perform pattern-matching on. Chunks too small
for pattern-matching will be put in the output FilteredOut table. If a
chunk splits a gap or elevation pattern in half,
<code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code> will attempt to detect this and report it
to the user as a ‘possible pattern-match continuity’ between
contig/genome chunks. Pattern-match continuity is detected when two
sequential chunks have a partial elevation/gap pattern going off the
right and left side of the chunks, respectively.</p>
</div>
<div class="section level4">
<h4 id="filtering">Filtering<a class="anchor" aria-label="anchor" href="#filtering"></a>
</h4>
<p>Contigs/chunks that are too short or have little to no read coverage
are filtered out prior to pattern-matching.
<code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code> filters out contigs/chunks that do not
have at least 10x coverage on a total of 5,000 bp across the whole
contig/chunk. The read coverage filtering was done in this way to avoid
filtering out long contigs/chunks with small elevations in read coverage
that might get removed if filtering was done with read coverage averages
or medians. Additionally, contigs/chunks less than 30,000 bp are
filtered out by default, however this can be changed with the
<code>minContigLength</code> parameter which can be set to a minimum of
25,000 bp. <strong>If you would like to reduce the size of your input
metagenome pileup file for</strong>
<strong><code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code>, consider pre-filtering your
assembly for contigs greater than</strong> <strong>25,000 bp prior to
read mapping!</strong></p>
</div>
<div class="section level4">
<h4 id="changing-pileup-windowsize">Changing pileup windowSize<a class="anchor" aria-label="anchor" href="#changing-pileup-windowsize"></a>
</h4>
<p>The input pileup files have 100 bp windows in which the mapped read
coverage is averaged over. <code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code> increases the
window size prior to pattern-matching by averaging the read coverages
over a value specified with <code>windowSize</code>. In many cases, read
coverage patterns don’t require the resolution that 100 bp windows
provide, however, starting with a 100 bp windowSize means the higher
resolution is available if needed. While users can use the 100 bp
<code>windowSize</code> for <code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code>, the
processing time will be increased <strong>significantly</strong> and
noisy data may interfere with pattern-matching. We find that the default
1,000 bp <code>windowSize</code> provides a nice balance between
processing time and read coverage pattern resolution. If you’d like more
resolution than the 1,000 bp <code>windowSize</code> provides, consider
dropping the <code>windowSize</code> to 500. If you’d like fine scale
read coverage resolution, consider viewing the contigs/genome with a
software like Integrative Genomics Viewer <a href="https://github.com/igvteam/igv" class="external-link">IGV</a>.</p>
</div>
<div class="section level4">
<h4 id="pattern-matching">Pattern-matching<a class="anchor" aria-label="anchor" href="#pattern-matching"></a>
</h4>
<p><code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code> detects read coverage patterns using a
2D pattern-matching algorithm. Several predefined patterns, described
below, are built using the specific length and read coverage values of
the contig/chunk being assessed. Patterns are translated across each
contig/chunk in 1,000 bp sliding windows and at each translation, a
pattern-match score is calculated by taking the mean absolute difference
of the read coverage and the pattern values. The smaller the
match-score, the better the pattern-match. After a pattern is fully
translated across a contig/chunk, certain aspects of the pattern are
changed (i.e. height, base, width) and translation is repeated. This
process of translation and pattern re-scaling is repeated until a large
number of pattern variations are tested. After pattern-matching is
complete, the pattern associated with the best match-score is used for
contig/chunk classification. Contigs/chunks are classified as
‘Elevation’, ‘Gap’, or ‘NoPattern’ during pattern-matching.</p>
<div class="section level5">
<h5 id="elevation-pattern">Elevation pattern:<a class="anchor" aria-label="anchor" href="#elevation-pattern"></a>
</h5>
<p>The ‘elevation’ class is defined by a ‘block’ pattern. During
pattern-matching, the height (max.), base (min.) and width are altered
and all pattern variations are translated across the contig/chunk. The
block width never gets smaller than 10,000 bp by default, however this
can be changed with the <code>minSize</code> parameter.</p>
<p><img src="ProActiveVignette_files/figure-html/unnamed-chunk-7-1.png" width="50%"></p>
</div>
<div class="section level5">
<h5 id="gap-pattern">Gap pattern:<a class="anchor" aria-label="anchor" href="#gap-pattern"></a>
</h5>
<p>The ‘gap’ class is essentially the reverse of the values used to
build the block pattern in the ‘elevation’ class. The same
pattern-matching steps (alteration of pattern max., min. and width
values and pattern translation) used for the elevation pattern are used
for the gap pattern.</p>
<p><img src="ProActiveVignette_files/figure-html/unnamed-chunk-8-1.png" width="50%"></p>
</div>
<div class="section level5">
<h5 id="elevationgap-pattern">Elevation/Gap pattern:<a class="anchor" aria-label="anchor" href="#elevationgap-pattern"></a>
</h5>
<p>Elevations and gaps that trail off one side of a contig/chunk are
hard to classify as the read coverage can be interpreted as a gap or
elevation depending on how you’re looking at it. We classify
contigs/chunk as ‘Gap’ if the elevated region is greater than 50% of the
length of the contig/chunk. Conversely, if the elevated region is less
than 50% of the contig/chunk length, the classification is
‘Elevation’.</p>
<p><img src="ProActiveVignette_files/figure-html/unnamed-chunk-9-1.png" width="50%"></p>
</div>
<div class="section level5">
<h5 id="nopattern">noPattern:<a class="anchor" aria-label="anchor" href="#nopattern"></a>
</h5>
<p>Since the best pattern-match for each contig/chunk is determined by
comparing match-scores amongst all pattern-variations from all pattern
classes, we needed a ‘negative control’ pattern to compare against. The
‘NoPattern’ ‘pattern’ serves as a negative control by matching to
contigs/chunks with no read coverage patterns. We made two NoPattern
patterns which consist of a horizontal line the same length as the
contig/chunk being assessed at the contig/chunk’s average and median
read coverage value. This pattern is not re-scaled or translated in any
way.</p>
<p><img src="ProActiveVignette_files/figure-html/unnamed-chunk-10-1.png" width="50%"></p>
</div>
</div>
<div class="section level4">
<h4 id="calculating-elevation-ratios">Calculating elevation ratios<a class="anchor" aria-label="anchor" href="#calculating-elevation-ratios"></a>
</h4>
<p>Every gap and elevation classification receives an ‘elevation ratio’
value which is simply the pattern-match’s maximum value divided by the
minimum value. For Elevation classifications, you can think of the
elevation ratio as how many times greater the read coverage of the
elevated region is compare to the non-elevated region’. Conversely, for
Gap classifications, the elevation ratio is how many times less the read
coverage of the gap region is compared to the non-gapped region.</p>
</div>
<div class="section level4">
<h4 id="extracting-gene-annotations-in-elevatedgapped-regions">Extracting gene annotations in elevated/gapped regions<a class="anchor" aria-label="anchor" href="#extracting-gene-annotations-in-elevatedgapped-regions"></a>
</h4>
<p>If a .gff file is provided, then <code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code> will
extract the gene annotations found within the gapped and elevated
pattern-match regions and provide them to the user in an output table
(GeneAnnotTable). An additional column will be added with the
classification information (Gap or Elevation) associated with the gene
annotations. If the input .gff file contains a gene ‘product’ field in
the attributes column (9th column in the dataframe), then
<code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code> will extract the product information into
a separate column for easy visualization and filtering of annotations of
interest.</p>
</div>
</div>
<div class="section level3">
<h3 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h3>
<p><strong>Default arguments in metagenome mode:</strong></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ProActiveOutputMetagenome</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ProActiveDetect.html">ProActiveDetect</a></span><span class="op">(</span></span>
<span>  pileup <span class="op">=</span> <span class="va">sampleMetagenomePileup</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"metagenome"</span>,</span>
<span>  gffTSV <span class="op">=</span> <span class="va">sampleMetagenomegffTSV</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Preparing input file for pattern-matching...</span></span>
<span><span class="co">#&gt; Starting pattern-matching...</span></span>
<span><span class="co">#&gt; A quarter of the way done with pattern-matching</span></span>
<span><span class="co">#&gt; Half of the way done with pattern-matching</span></span>
<span><span class="co">#&gt; Almost done with pattern-matching!</span></span>
<span><span class="co">#&gt; Summarizing pattern-matching results</span></span>
<span><span class="co">#&gt; Finding gene predictions in elevated or gapped regions of read coverage...</span></span>
<span><span class="co">#&gt; Finalizing output</span></span>
<span><span class="co">#&gt; Execution time: 1.99secs</span></span>
<span><span class="co">#&gt; 0 contigs were filtered out based on low read coverage</span></span>
<span><span class="co">#&gt; 0 contigs were filtered out based on length (&lt; minContigLength)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Elevation       Gap NoPattern </span></span>
<span><span class="co">#&gt;         3         3         1</span></span></code></pre></div>
<p><strong>Default arguments in genome mode:</strong></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ProActiveOutputGenome</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ProActiveDetect.html">ProActiveDetect</a></span><span class="op">(</span></span>
<span>  pileup <span class="op">=</span> <span class="va">sampleGenomePileup</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  gffTSV <span class="op">=</span> <span class="va">sampleGenomegffTSV</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Preparing input file for pattern-matching...</span></span>
<span><span class="co">#&gt; Starting pattern-matching...</span></span>
<span><span class="co">#&gt; A quarter of the way done with pattern-matching</span></span>
<span><span class="co">#&gt; Half of the way done with pattern-matching</span></span>
<span><span class="co">#&gt; Almost done with pattern-matching!</span></span>
<span><span class="co">#&gt; Summarizing pattern-matching results</span></span>
<span><span class="co">#&gt; Finding gene predictions in elevated or gapped regions of read coverage...</span></span>
<span><span class="co">#&gt; Finalizing output</span></span>
<span><span class="co">#&gt; Execution time: 30.27secs</span></span>
<span><span class="co">#&gt; 0 contigs were filtered out based on low read coverage</span></span>
<span><span class="co">#&gt; 0 contigs were filtered out based on length (&lt; minContigLength)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Elevation       Gap NoPattern </span></span>
<span><span class="co">#&gt;        25         3        21</span></span></code></pre></div>
<p><strong>Note that <code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code> <em>can</em> be run
without the gffTSV file!</strong></p>
</div>
<div class="section level3">
<h3 id="argumentsparameters">Arguments/parameters<a class="anchor" aria-label="anchor" href="#argumentsparameters"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ProActiveDetect.html">ProActiveDetect</a></span><span class="op">(</span></span>
<span>  <span class="va">pileup</span>,</span>
<span>  <span class="va">mode</span>,</span>
<span>  <span class="va">gffTSV</span>,</span>
<span>  windowSize <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  minSize <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  maxSize <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  minContigLength <span class="op">=</span> <span class="fl">30000</span>,</span>
<span>  chunkSize <span class="op">=</span> <span class="fl">50000</span>,</span>
<span>  chunkContigs <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  IncludeNoPatterns <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">saveFilesTo</span></span>
<span><span class="op">)</span></span></code></pre></div>
<ul>
<li>
<strong><code>pileup</code></strong>: A .txt file containing mapped
sequencing read coverages averaged over 100 bp windows/bins.</li>
<li>
<strong><code>mode</code></strong>: Either “genome” or
“metagenome”.</li>
<li>
<strong><code>gffTSV</code></strong>: Optional, a .gff file (TSV)
containing gene annotations associated with the .fasta file used to
generate the pileup.</li>
<li>
<strong><code>windowSize</code></strong>: The number of basepairs to
average read coverage values over. Options are 100, 200, 500, 1000 ONLY.
Default is 1000.</li>
<li>
<strong><code>minSize</code></strong>: The minimum size (in bp) of
elevation or gap patterns. Default is 10000.</li>
<li>
<strong><code>maxSize</code></strong>: The maximum size (in bp) of
elevation or gap patterns. Default is NA (i.e. no maximum).</li>
<li>
<strong><code>minContigLength</code></strong>: The minimum
contig/chunk size (in bp) to perform pattern-matching on. Default is
25000.</li>
<li>
<strong><code>chunkSize</code></strong>: If
<code>mode = "genome"</code> OR if <code>mode = "metagenome"</code> and
<code>chunkContigs = TRUE</code>, chunk the genome or contigs,
respectively, into smaller subsets for pattern-matching.
<code>chunkSize</code> determines the size (in bp) of each ‘chunk’.
Default is 100000.</li>
<li>
<strong><code>chunkContigs</code></strong>: TRUE or FALSE, If TRUE
and <code>mode = "metagenome"</code>, contigs longer than the
<code>chunkSize</code> will be ‘chunked’ into smaller subsets and
pattern-matching will be performed on each subset. Default is
FALSE.</li>
<li>
<strong><code>IncludeNoPatterns</code></strong>: TRUE or FALSE, If
TRUE the noPattern pattern-matches will be included in the
PatternMatches output list. If you would like to visualize the read
coverage of noPattern classifications in
<code><a href="../reference/plotProActiveResults.html">plotProActiveResults()</a></code>, this should be set to TRUE.</li>
<li>
<strong><code>verbose</code></strong>: TRUE or FALSE. Print progress
messages to console. Default is TRUE.</li>
<li>
<strong><code>saveFilesTo</code></strong>: Optional, Provide a path
to the directory you wish to save output to. A folder will be made
within the provided directory to store results.</li>
</ul>
</div>
<div class="section level3">
<h3 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h3>
<p>The output of <code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code> is a list containing six
objects:</p>
<ol style="list-style-type: decimal">
<li>SummaryTable: A table containing all pattern-matching
classifications</li>
<li>CleanSummaryTable: A table containing only Gap and Elevation
pattern-match classifications (i.e. noPattern classifications
removed)</li>
<li>PatternMatches: A list object containing information needed to
visualize the pattern-matches in
<code><a href="../reference/plotProActiveResults.html">plotProActiveResults()</a></code>
</li>
<li>FilteredOut: A table containing contigs/chunks that were filtered
out for being too small or having too low read coverage</li>
<li>Arguments: A list object containing arguments used for
pattern-matching (windowSize, mode, chunkSize, chunkContigs)</li>
<li>GeneAnnotTable: A table containing gene annotations associated with
elevated or gapped regions in pattern-matches</li>
</ol>
<p>Save the desired list item to a new variable using its associated
name.</p>
<p><strong>Metagenome results summary table:</strong></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MetagenomeCleanSummaryTable</span> <span class="op">&lt;-</span> <span class="va">ProActiveOutputMetagenome</span><span class="op">$</span><span class="va">CleanSummaryTable</span></span></code></pre></div>
<table class="table table" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
refName
</th>
<th style="text-align:left;">
classification
</th>
<th style="text-align:right;">
elevRatio
</th>
<th style="text-align:right;">
startPos
</th>
<th style="text-align:right;">
endPos
</th>
<th style="text-align:right;">
matchSize
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
NODE_1911
</td>
<td style="text-align:left;">
Elevation
</td>
<td style="text-align:right;">
3.349296
</td>
<td style="text-align:right;">
1000
</td>
<td style="text-align:right;">
17000
</td>
<td style="text-align:right;">
16000
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1583
</td>
<td style="text-align:left;">
Elevation
</td>
<td style="text-align:right;">
2.450013
</td>
<td style="text-align:right;">
42000
</td>
<td style="text-align:right;">
51000
</td>
<td style="text-align:right;">
9000
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1884
</td>
<td style="text-align:left;">
Gap
</td>
<td style="text-align:right;">
3.319514
</td>
<td style="text-align:right;">
36000
</td>
<td style="text-align:right;">
44000
</td>
<td style="text-align:right;">
8000
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1255
</td>
<td style="text-align:left;">
Gap
</td>
<td style="text-align:right;">
5.318072
</td>
<td style="text-align:right;">
1000
</td>
<td style="text-align:right;">
20000
</td>
<td style="text-align:right;">
19000
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_368
</td>
<td style="text-align:left;">
Gap
</td>
<td style="text-align:right;">
2.690172
</td>
<td style="text-align:right;">
26000
</td>
<td style="text-align:right;">
56000
</td>
<td style="text-align:right;">
30000
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_617
</td>
<td style="text-align:left;">
Elevation
</td>
<td style="text-align:right;">
1.784556
</td>
<td style="text-align:right;">
34000
</td>
<td style="text-align:right;">
82000
</td>
<td style="text-align:right;">
48000
</td>
</tr>
</tbody>
</table>
<p><strong>Subset of genome results summary table:</strong></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GenomeCleanSummaryTable</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ProActiveOutputGenome</span><span class="op">$</span><span class="va">CleanSummaryTable</span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
refName
</th>
<th style="text-align:left;">
classification
</th>
<th style="text-align:right;">
elevRatio
</th>
<th style="text-align:right;">
startPos
</th>
<th style="text-align:right;">
endPos
</th>
<th style="text-align:right;">
matchSize
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
3
</td>
<td style="text-align:left;">
NC_003197.2_chunk_3
</td>
<td style="text-align:left;">
Gap
</td>
<td style="text-align:right;">
1.609512
</td>
<td style="text-align:right;">
283000
</td>
<td style="text-align:right;">
300000
</td>
<td style="text-align:right;">
17000
</td>
</tr>
<tr>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
NC_003197.2_chunk_4
</td>
<td style="text-align:left;">
Elevation
</td>
<td style="text-align:right;">
1.351684
</td>
<td style="text-align:right;">
339000
</td>
<td style="text-align:right;">
351000
</td>
<td style="text-align:right;">
12000
</td>
</tr>
<tr>
<td style="text-align:left;">
7
</td>
<td style="text-align:left;">
NC_003197.2_chunk_7
</td>
<td style="text-align:left;">
Elevation
</td>
<td style="text-align:right;">
1.343521
</td>
<td style="text-align:right;">
692000
</td>
<td style="text-align:right;">
700000
</td>
<td style="text-align:right;">
8000
</td>
</tr>
<tr>
<td style="text-align:left;">
8
</td>
<td style="text-align:left;">
NC_003197.2_chunk_8
</td>
<td style="text-align:left;">
Elevation
</td>
<td style="text-align:right;">
1.723227
</td>
<td style="text-align:right;">
784000
</td>
<td style="text-align:right;">
794000
</td>
<td style="text-align:right;">
10000
</td>
</tr>
<tr>
<td style="text-align:left;">
10
</td>
<td style="text-align:left;">
NC_003197.2_chunk_10
</td>
<td style="text-align:left;">
Elevation
</td>
<td style="text-align:right;">
1.887571
</td>
<td style="text-align:right;">
964000
</td>
<td style="text-align:right;">
1000000
</td>
<td style="text-align:right;">
36000
</td>
</tr>
<tr>
<td style="text-align:left;">
12
</td>
<td style="text-align:left;">
NC_003197.2_chunk_12
</td>
<td style="text-align:left;">
Elevation
</td>
<td style="text-align:right;">
1.293144
</td>
<td style="text-align:right;">
1134000
</td>
<td style="text-align:right;">
1144000
</td>
<td style="text-align:right;">
10000
</td>
</tr>
</tbody>
</table>
<p><strong>Subset of GeneAnnotTable for metagenome results:</strong></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MetagenomeResultsGenePredictTable</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ProActiveOutputMetagenome</span><span class="op">$</span><span class="va">GeneAnnotTable</span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
seqid
</th>
<th style="text-align:left;">
source
</th>
<th style="text-align:left;">
type
</th>
<th style="text-align:right;">
start
</th>
<th style="text-align:right;">
end
</th>
<th style="text-align:left;">
score
</th>
<th style="text-align:left;">
strand
</th>
<th style="text-align:left;">
phase
</th>
<th style="text-align:left;">
attributes
</th>
<th style="text-align:left;">
geneproduct
</th>
<th style="text-align:left;">
Classification
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
NODE_1911
</td>
<td style="text-align:left;">
Prodigal:002006
</td>
<td style="text-align:left;">
CDS
</td>
<td style="text-align:right;">
318
</td>
<td style="text-align:right;">
1166
</td>
<td style="text-align:left;">
.
</td>
<td style="text-align:left;">
<ul><li></ul>
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
ID=NJKKNKEE_164175;inference=ab initio
prediction:Prodigal:002006;locus_tag=NJKKNKEE_164175;product=hypothetical
protein
</td>
<td style="text-align:left;">
hypothetical protein
</td>
<td style="text-align:left;">
Elevation
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1911
</td>
<td style="text-align:left;">
Prodigal:002006
</td>
<td style="text-align:left;">
CDS
</td>
<td style="text-align:right;">
1198
</td>
<td style="text-align:right;">
1938
</td>
<td style="text-align:left;">
.
</td>
<td style="text-align:left;">
<ul><li></ul>
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
ID=NJKKNKEE_164176;inference=ab initio
prediction:Prodigal:002006;locus_tag=NJKKNKEE_164176;product=hypothetical
protein
</td>
<td style="text-align:left;">
hypothetical protein
</td>
<td style="text-align:left;">
Elevation
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1911
</td>
<td style="text-align:left;">
Prodigal:002006
</td>
<td style="text-align:left;">
CDS
</td>
<td style="text-align:right;">
1938
</td>
<td style="text-align:right;">
2582
</td>
<td style="text-align:left;">
.
</td>
<td style="text-align:left;">
<ul><li></ul>
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
ID=NJKKNKEE_164177;inference=ab initio
prediction:Prodigal:002006;locus_tag=NJKKNKEE_164177;product=hypothetical
protein
</td>
<td style="text-align:left;">
hypothetical protein
</td>
<td style="text-align:left;">
Elevation
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1911
</td>
<td style="text-align:left;">
Prodigal:002006
</td>
<td style="text-align:left;">
CDS
</td>
<td style="text-align:right;">
2722
</td>
<td style="text-align:right;">
3561
</td>
<td style="text-align:left;">
.
</td>
<td style="text-align:left;">
<ul><li></ul>
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
ID=NJKKNKEE_164178;inference=ab initio
prediction:Prodigal:002006;locus_tag=NJKKNKEE_164178;product=hypothetical
protein
</td>
<td style="text-align:left;">
hypothetical protein
</td>
<td style="text-align:left;">
Elevation
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1911
</td>
<td style="text-align:left;">
Prodigal:002006
</td>
<td style="text-align:left;">
CDS
</td>
<td style="text-align:right;">
3671
</td>
<td style="text-align:right;">
4063
</td>
<td style="text-align:left;">
.
</td>
<td style="text-align:left;">
<ul><li></ul>
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
ID=NJKKNKEE_164179;inference=ab initio
prediction:Prodigal:002006;locus_tag=NJKKNKEE_164179;product=hypothetical
protein
</td>
<td style="text-align:left;">
hypothetical protein
</td>
<td style="text-align:left;">
Elevation
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1911
</td>
<td style="text-align:left;">
Prodigal:002006
</td>
<td style="text-align:left;">
CDS
</td>
<td style="text-align:right;">
4128
</td>
<td style="text-align:right;">
4670
</td>
<td style="text-align:left;">
.
</td>
<td style="text-align:left;">
<ul><li></ul>
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
ID=NJKKNKEE_164180;eC_number=1.11.1.1;Name=rbr1_24;db_xref=COG:COG1592;gene=rbr1_24;inference=ab
initio prediction:Prodigal:002006,similar to AA
sequence:UniProtKB:Q97FZ9;locus_tag=NJKKNKEE_164180;product=Rubrerythrin-1
</td>
<td style="text-align:left;">
Rubrerythrin-1
</td>
<td style="text-align:left;">
Elevation
</td>
</tr>
</tbody>
</table>
<p><strong>Subset of GeneAnnotTable for genome results:</strong></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GenomeResultsGenePredictTable</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ProActiveOutputGenome</span><span class="op">$</span><span class="va">GeneAnnotTable</span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
seqid
</th>
<th style="text-align:left;">
source
</th>
<th style="text-align:left;">
type
</th>
<th style="text-align:right;">
start
</th>
<th style="text-align:right;">
end
</th>
<th style="text-align:left;">
score
</th>
<th style="text-align:left;">
strand
</th>
<th style="text-align:left;">
phase
</th>
<th style="text-align:left;">
attributes
</th>
<th style="text-align:left;">
geneproduct
</th>
<th style="text-align:left;">
Classification
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
NC_003197.2_chunk_3
</td>
<td style="text-align:left;">
Prodigal:002006
</td>
<td style="text-align:left;">
CDS
</td>
<td style="text-align:right;">
282468
</td>
<td style="text-align:right;">
284186
</td>
<td style="text-align:left;">
.
</td>
<td style="text-align:left;">
<ul><li></ul>
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
ID=LFLNNMPD_00240;eC_number=6.1.1.15;Name=proS;db_xref=COG:COG0442;gene=proS;inference=ab
initio prediction:Prodigal:002006,similar to AA
sequence:UniProtKB:P16659;locus_tag=LFLNNMPD_00240;product=Proline–tRNA
ligase
</td>
<td style="text-align:left;">
Proline–tRNA ligase
</td>
<td style="text-align:left;">
Gap
</td>
</tr>
<tr>
<td style="text-align:left;">
NC_003197.2_chunk_3
</td>
<td style="text-align:left;">
Prodigal:002006
</td>
<td style="text-align:left;">
CDS
</td>
<td style="text-align:right;">
284297
</td>
<td style="text-align:right;">
285004
</td>
<td style="text-align:left;">
.
</td>
<td style="text-align:left;">
<ul><li></ul>
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
ID=LFLNNMPD_00241;eC_number=2.1.1.-;Name=trmO;db_xref=COG:COG1720;gene=trmO;inference=ab
initio prediction:Prodigal:002006,similar to AA
sequence:UniProtKB:P28634;locus_tag=LFLNNMPD_00241;product=tRNA
(adenine(37)-N6)-methyltransferase
</td>
<td style="text-align:left;">
tRNA (adenine(37)-N6)-methyltransferase
</td>
<td style="text-align:left;">
Gap
</td>
</tr>
<tr>
<td style="text-align:left;">
NC_003197.2_chunk_3
</td>
<td style="text-align:left;">
Prodigal:002006
</td>
<td style="text-align:left;">
CDS
</td>
<td style="text-align:right;">
285001
</td>
<td style="text-align:right;">
285405
</td>
<td style="text-align:left;">
.
</td>
<td style="text-align:left;">
<ul><li></ul>
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
ID=LFLNNMPD_00242;Name=rcsF;gene=rcsF;inference=ab initio
prediction:Prodigal:002006,similar to AA
sequence:UniProtKB:P69411;locus_tag=LFLNNMPD_00242;product=Outer
membrane lipoprotein RcsF
</td>
<td style="text-align:left;">
Outer membrane lipoprotein RcsF
</td>
<td style="text-align:left;">
Gap
</td>
</tr>
<tr>
<td style="text-align:left;">
NC_003197.2_chunk_3
</td>
<td style="text-align:left;">
Prodigal:002006
</td>
<td style="text-align:left;">
CDS
</td>
<td style="text-align:right;">
285524
</td>
<td style="text-align:right;">
286339
</td>
<td style="text-align:left;">
.
</td>
<td style="text-align:left;">
<ul><li></ul>
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
ID=LFLNNMPD_00243;Name=metQ_1;db_xref=COG:COG1464;gene=metQ_1;inference=ab
initio prediction:Prodigal:002006,similar to AA
sequence:UniProtKB:P28635;locus_tag=LFLNNMPD_00243;product=D-methionine-binding
lipoprotein MetQ
</td>
<td style="text-align:left;">
D-methionine-binding lipoprotein MetQ
</td>
<td style="text-align:left;">
Gap
</td>
</tr>
<tr>
<td style="text-align:left;">
NC_003197.2_chunk_3
</td>
<td style="text-align:left;">
Prodigal:002006
</td>
<td style="text-align:left;">
CDS
</td>
<td style="text-align:right;">
286378
</td>
<td style="text-align:right;">
287031
</td>
<td style="text-align:left;">
.
</td>
<td style="text-align:left;">
<ul><li></ul>
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
ID=LFLNNMPD_00244;Name=metI;db_xref=COG:COG2011;gene=metI;inference=ab
initio prediction:Prodigal:002006,similar to AA
sequence:UniProtKB:P31547;locus_tag=LFLNNMPD_00244;product=D-methionine
transport system permease protein MetI
</td>
<td style="text-align:left;">
D-methionine transport system permease protein MetI
</td>
<td style="text-align:left;">
Gap
</td>
</tr>
<tr>
<td style="text-align:left;">
NC_003197.2_chunk_3
</td>
<td style="text-align:left;">
Prodigal:002006
</td>
<td style="text-align:left;">
CDS
</td>
<td style="text-align:right;">
287024
</td>
<td style="text-align:right;">
288055
</td>
<td style="text-align:left;">
.
</td>
<td style="text-align:left;">
<ul><li></ul>
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
ID=LFLNNMPD_00245;eC_number=3.6.3.-;Name=metN_1;db_xref=COG:COG1135;gene=metN_1;inference=ab
initio prediction:Prodigal:002006,similar to AA
sequence:UniProtKB:P30750;locus_tag=LFLNNMPD_00245;product=Methionine
import ATP-binding protein MetN
</td>
<td style="text-align:left;">
Methionine import ATP-binding protein MetN
</td>
<td style="text-align:left;">
Gap
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="plotproactiveresults">plotProActiveResults()<a class="anchor" aria-label="anchor" href="#plotproactiveresults"></a>
</h2>
<p><code><a href="../reference/plotProActiveResults.html">plotProActiveResults()</a></code> allows users to visualize both
the read coverage and the pattern-match associated with each Gap or
Elevation classification.</p>
<div class="section level3">
<h3 id="function-components-1">Function components<a class="anchor" aria-label="anchor" href="#function-components-1"></a>
</h3>
<div class="section level4">
<h4 id="re-building-pattern-matches">Re-building pattern-matches<a class="anchor" aria-label="anchor" href="#re-building-pattern-matches"></a>
</h4>
<p>The <code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code> output contains information needed
to re-build each pattern-match used for classification. To re-build a
complete pattern-match for visualization,
<code><a href="../reference/plotProActiveResults.html">plotProActiveResults()</a></code> uses the pattern-match’s minimum and
maximum values and the start and stop positions.</p>
</div>
<div class="section level4">
<h4 id="plotting-read-coverage-and-associated-pattern-matches">Plotting read coverage and associated pattern-matches<a class="anchor" aria-label="anchor" href="#plotting-read-coverage-and-associated-pattern-matches"></a>
</h4>
<p>By default, the read coverage is plotted for each contig/chunk
classified as having a Gap or Elevation in read coverage. If you wish to
see the read coverage for noPattern classifications, be sure to set
<code>IncludeNoPatterns = TRUE</code> when running
<code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code>. The pattern-match associated with each
classification is overlaid on the coverage plot.</p>
</div>
</div>
<div class="section level3">
<h3 id="usage-1">Usage<a class="anchor" aria-label="anchor" href="#usage-1"></a>
</h3>
<p><strong>Default arguments:</strong></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MetagenomeResultsPlots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotProActiveResults.html">plotProActiveResults</a></span><span class="op">(</span></span>
<span>  pileup <span class="op">=</span> <span class="va">sampleMetagenomePileup</span>,</span>
<span>  ProActiveResults <span class="op">=</span> <span class="va">ProActiveOutputMetagenome</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">GenomeResultsPlots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotProActiveResults.html">plotProActiveResults</a></span><span class="op">(</span></span>
<span>  pileup <span class="op">=</span> <span class="va">sampleGenomePileup</span>,</span>
<span>  ProActiveResults <span class="op">=</span> <span class="va">ProActiveOutputGenome</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>Note</strong>- There is no need to set ‘genome’ or
‘metagenome’ mode. <code><a href="../reference/plotProActiveResults.html">plotProActiveResults()</a></code> will get this
information from the <code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code> output.</p>
</div>
<div class="section level3">
<h3 id="argumentsparameters-1">Arguments/parameters<a class="anchor" aria-label="anchor" href="#argumentsparameters-1"></a>
</h3>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotProActiveResults.html">plotProActiveResults</a></span><span class="op">(</span><span class="va">pileup</span>,</span>
<span>  <span class="va">ProActiveResults</span>,</span>
<span>  <span class="va">elevFilter</span>,</span>
<span>  <span class="va">saveFilesTo</span></span>
<span><span class="op">)</span></span></code></pre></div>
<ul>
<li>
<strong><code>pileup</code></strong>: A .txt file containing mapped
sequencing read coverages averaged over 100 bp windows/bins.</li>
<li>
<strong><code>ProActiveResults</code></strong>: The output from
<code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code>.</li>
<li>
<strong><code>elevFilter</code></strong>: Optional, only plot
results with pattern-matches that achieved an elevation ratio (max/min)
greater than the specified value. Default is no filter.</li>
<li>
<strong><code>saveFilesTo</code></strong>: Optional, Provide a path
to the directory you wish to save output to. A folder will be made
within the provided directory to store results.</li>
</ul>
</div>
<div class="section level3">
<h3 id="output-1">Output<a class="anchor" aria-label="anchor" href="#output-1"></a>
</h3>
<p>The output of <code><a href="../reference/plotProActiveResults.html">plotProActiveResults()</a></code> is a list of ggplot
objects.</p>
<div class="section level4">
<h4 id="view-select-metagenome-plots">View select metagenome plots<a class="anchor" aria-label="anchor" href="#view-select-metagenome-plots"></a>
</h4>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MetagenomeResultsPlots</span><span class="op">$</span><span class="va">NODE_1884</span></span></code></pre></div>
<p><img src="ProActiveVignette_files/figure-html/unnamed-chunk-24-1.png" width="100%"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MetagenomeResultsPlots</span><span class="op">$</span><span class="va">NODE_368</span></span></code></pre></div>
<p><img src="ProActiveVignette_files/figure-html/unnamed-chunk-24-2.png" width="100%"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MetagenomeResultsPlots</span><span class="op">$</span><span class="va">NODE_617</span></span></code></pre></div>
<p><img src="ProActiveVignette_files/figure-html/unnamed-chunk-24-3.png" width="100%"></p>
</div>
<div class="section level4">
<h4 id="view-select-genome-plots">View select genome plots<a class="anchor" aria-label="anchor" href="#view-select-genome-plots"></a>
</h4>
<p><strong>Notice the ‘chunk’ information in the plot
titles</strong></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GenomeResultsPlots</span><span class="op">$</span><span class="va">NC_003197.2_chunk_36</span></span></code></pre></div>
<p><img src="ProActiveVignette_files/figure-html/unnamed-chunk-25-1.png" width="100%"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GenomeResultsPlots</span><span class="op">$</span><span class="va">NC_003197.2_chunk_8</span></span></code></pre></div>
<p><img src="ProActiveVignette_files/figure-html/unnamed-chunk-25-2.png" width="100%"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="geneannotationsearch">geneAnnotationSearch()<a class="anchor" aria-label="anchor" href="#geneannotationsearch"></a>
</h2>
<p><code><a href="../reference/geneAnnotationSearch.html">geneAnnotationSearch()</a></code> helps users explore gene
annotations of interest in and around detected gaps and elevations in
read coverage.</p>
<div class="section level3">
<h3 id="function-components-2">Function components<a class="anchor" aria-label="anchor" href="#function-components-2"></a>
</h3>
<div class="section level4">
<h4 id="search-for-gene-annotations">Search for gene annotations<a class="anchor" aria-label="anchor" href="#search-for-gene-annotations"></a>
</h4>
<p><code><a href="../reference/geneAnnotationSearch.html">geneAnnotationSearch()</a></code> utilizes a .gff file and the
pattern-matching results from <code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code> to locate
gene annotations that match provided <code>keyWords</code>. The .gff
file should be in the same format described previously in the Input
Files section of this vignette. First, the information associated with
the gene or gene product (depending on what the user selects for the
<code>geneOrProduct</code> parameter) is extracted from the attributes
column of the .gff file. Then, the .gff file is subset to include only
the annotations associated with the contig/chunk being assessed. From
here, the search can vary quite a bit depending on the parameters the
user selects for the <code>inGapOrElev</code> and <code>bpRange</code>
parameters. If <code>inGapOrElev = FALSE</code> (the default), then gene
annotations located anywhere on the contig/chunk that match one or more
of the provided <code>keyWords</code> will be visualized. If
<code>inGapOrElev = TRUE</code>, then only gene annotations within the
gap/elevation region of the pattern-match will be searched for matches
to the provided <code>keyWords</code>. The <code>bpRange</code>
parameter can be used if <code>inGapOrElev = TRUE</code> and allows the
search range to be extended a specified number of base pairs to the left
and right of the gap/elevation pattern-match borders. Gene annotation
are included in the search if the end of the open reading frame (defined
by the ‘end’ values in the .gff file) falls within the search
region.</p>
</div>
<div class="section level4">
<h4 id="plot-gene-annotation-locations">Plot gene annotation locations<a class="anchor" aria-label="anchor" href="#plot-gene-annotation-locations"></a>
</h4>
<p>The read coverage and locations of gene annotations that match the
provided <code>keyWords</code> are visualized for each contig/chunk with
matches. The read coverage is plotted using a 100 bp windowSize to allow
for greater resolution of read coverage patterns and gene annotation
locations. The borders of the elevated/gapped regions of read coverage
detected by <code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code> are marked on the plot with
orange vertical lines. If <code>inElevOrGap = TRUE</code> and
<code>bpRange</code> is set to a non-zero value, then the extended
search range outside the gap/elevation borders are marked on the plot
with orange dashed vertical lines. The matching gene annotation
locations are marked on the plot with black vertical lines at the start
position of the associated open reading frames.</p>
<p><strong>Note</strong>- The pattern-matching used to identify the
borders of elevated and gapped regions of read coverage was
<em>likely</em> performed using a <code>windowSize</code> larger than
100 bp in <code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code>. This means that the locations
of the borders may not perfectly translate to the borders of gaps and
elevations at 100 bp resolution.</p>
</div>
</div>
<div class="section level3">
<h3 id="usage-2">Usage<a class="anchor" aria-label="anchor" href="#usage-2"></a>
</h3>
<p><strong>Default arguments:</strong></p>
<p>With defaults, all contigs/chunks classified as having a gap or
elevation in read coverage are searched for gene annotations that match
any of the provided keywords. The entire contig/chunk is searched, not
just the gapped or elevated region.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MetagenomeGeneMatches</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geneAnnotationSearch.html">geneAnnotationSearch</a></span><span class="op">(</span></span>
<span>  ProActiveResults <span class="op">=</span> <span class="va">ProActiveOutputMetagenome</span>, </span>
<span>  pileup <span class="op">=</span> <span class="va">sampleMetagenomePileup</span>, </span>
<span>  gffTSV <span class="op">=</span> <span class="va">sampleMetagenomegffTSV</span>, </span>
<span>  geneOrProduct <span class="op">=</span> <span class="st">"product"</span>, </span>
<span>  keyWords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"transport"</span>, <span class="st">"chemotaxis"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; Cleaning gff file...</span></span>
<span><span class="co">#&gt; Cleaning pileup file...</span></span>
<span><span class="co">#&gt; Searching for matching annotations...</span></span>
<span><span class="co">#&gt; 3 contigs/chunks have gene annotations that match one or more of the provided keyWords</span></span></code></pre></div>
<p><strong>Non-default arguments</strong></p>
<p>With the following parameters/arguments, all classified
contigs/chunks are searched for gene annotations that match the provided
keywords (same as default), BUT with the use of
<code>inGapOrElev = TRUE</code>, only the gapped or elevated region is
searched for matching annotations. Additionally, the use of
<code>bpRange = 5000</code> means that the search region is extended
5,000 bp from the left and right of the gapped or elevated region.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GenomeGeneMatches</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geneAnnotationSearch.html">geneAnnotationSearch</a></span><span class="op">(</span></span>
<span>  ProActiveResults <span class="op">=</span> <span class="va">ProActiveOutputGenome</span>,</span>
<span>  pileup <span class="op">=</span>  <span class="va">sampleGenomePileup</span>, </span>
<span>  gffTSV <span class="op">=</span> <span class="va">sampleGenomegffTSV</span>, </span>
<span>  geneOrProduct <span class="op">=</span> <span class="st">"product"</span>, </span>
<span>  keyWords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ribosomal"</span><span class="op">)</span>, </span>
<span>  inGapOrElev <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>  bpRange <span class="op">=</span> <span class="fl">5000</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; Cleaning gff file...</span></span>
<span><span class="co">#&gt; Cleaning pileup file...</span></span>
<span><span class="co">#&gt; Searching for matching annotations...</span></span>
<span><span class="co">#&gt; 8 contigs/chunks have gene annotations that match one or more of the provided keyWords</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="argumentsparameters-2">Arguments/parameters<a class="anchor" aria-label="anchor" href="#argumentsparameters-2"></a>
</h3>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/geneAnnotationSearch.html">geneAnnotationSearch</a></span><span class="op">(</span></span>
<span>  <span class="va">ProActiveResults</span>, </span>
<span>  <span class="va">pileup</span>, </span>
<span>  <span class="va">gffTSV</span>,</span>
<span>  <span class="va">geneOrProduct</span>, </span>
<span>  <span class="va">keyWords</span>, </span>
<span>  inGapOrElev <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  bpRange <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  <span class="va">elevFilter</span>, </span>
<span>  <span class="va">saveFilesTo</span>, </span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span> </span></code></pre></div>
<ul>
<li>
<strong><code>ProActiveResults</code></strong>: The output from
<code><a href="../reference/ProActiveDetect.html">ProActiveDetect()</a></code>.</li>
<li>
<strong><code>pileup</code></strong>: A .txt file containing mapped
sequencing read coverages averaged over 100 bp windows/bins.</li>
<li>
<strong><code>gffTSV</code></strong>: A .gff file (TSV) containing
gene annotations associated with the .fasta file used to generate the
pileup.</li>
<li>
<strong><code>geneOrProduct</code></strong>: “gene” or “product”.
Search for keyWords associated with genes or gene products.</li>
<li>
<strong><code>keyWords</code></strong>: The keyWord(s) to search
for. Case independent. Searches will return the string #’ that contains
the matching keyWord. KeyWord(s) must be in quotes, comma-separated, and
surrounded by #’ c() i.e( c(“antibiotic”, “resistance”, “drug”) )</li>
<li>
<strong><code>inGapOrElev</code></strong>: TRUE or FALSE. If TRUE,
only search for gene-annotations in #’ the gap/elevation region of the
pattern-match. Default is FALSE (i.e search the #’ entire contig/chunk
for the gene annotation key-words)</li>
<li>
<strong><code>bpRange</code></strong>: If
<code>inGapOrElev = TRUE</code>, the user may specify the region (in
base pairs) that should #’ be searched to the left and right of the
gap/elevation region. Default is 0.</li>
<li>
<strong><code>elevFilter</code></strong>: Optional, only plot
results with pattern-matches that achieved an elevation ratio (max/min)
greater than the specified value. Default is no filter.</li>
<li>
<strong><code>saveFilesTo</code></strong>: Optional, Provide a path
to the directory you wish to save output to. A folder will be made
within the provided directory to store results.</li>
<li>
<strong><code>verbose</code></strong>: TRUE or FALSE. Print progress
messages to console. Default is TRUE.</li>
</ul>
</div>
<div class="section level3">
<h3 id="output-2">Output<a class="anchor" aria-label="anchor" href="#output-2"></a>
</h3>
<p>The output of <code><a href="../reference/geneAnnotationSearch.html">geneAnnotationSearch()</a></code> is a list of ggplot
objects.</p>
<p>Default search parameters:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MetagenomeGeneMatches</span><span class="op">$</span><span class="va">NODE_617</span></span></code></pre></div>
<p><img src="ProActiveVignette_files/figure-html/unnamed-chunk-29-1.png" width="100%"></p>
<p>Non-default search parameters (use of <code>inGapOrElev</code> = TRUE
and <code>bpRange</code> = 5000)</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GenomeGeneMatches</span><span class="op">$</span><span class="va">NC_003197.2_chunk_3</span></span></code></pre></div>
<p><img src="ProActiveVignette_files/figure-html/unnamed-chunk-30-1.png" width="100%"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GenomeGeneMatches</span><span class="op">$</span><span class="va">NC_003197.2_chunk_36</span></span></code></pre></div>
<p><img src="ProActiveVignette_files/figure-html/unnamed-chunk-30-2.png" width="100%"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.1 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] dplyr_1.1.4          stringr_1.5.1        ggplot2_3.5.1       </span></span>
<span><span class="co">#&gt; [4] kableExtra_1.4.0     ProActive_0.1.0.9000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] gtable_0.3.6      jsonlite_1.8.9    compiler_4.4.2    tidyselect_1.2.1 </span></span>
<span><span class="co">#&gt;  [5] xml2_1.3.6        jquerylib_0.1.4   systemfonts_1.2.0 scales_1.3.0     </span></span>
<span><span class="co">#&gt;  [9] textshaping_0.4.1 yaml_2.3.10       fastmap_1.2.0     R6_2.5.1         </span></span>
<span><span class="co">#&gt; [13] labeling_0.4.3    generics_0.1.3    knitr_1.49        tibble_3.2.1     </span></span>
<span><span class="co">#&gt; [17] desc_1.4.3        munsell_0.5.1     svglite_2.1.3     bslib_0.8.0      </span></span>
<span><span class="co">#&gt; [21] pillar_1.10.1     rlang_1.1.5       cachem_1.1.0      stringi_1.8.4    </span></span>
<span><span class="co">#&gt; [25] xfun_0.50         fs_1.6.5          sass_0.4.9        viridisLite_0.4.2</span></span>
<span><span class="co">#&gt; [29] cli_3.6.3         withr_3.0.2       pkgdown_2.1.1     magrittr_2.0.3   </span></span>
<span><span class="co">#&gt; [33] digest_0.6.37     grid_4.4.2        rstudioapi_0.17.1 lifecycle_1.0.4  </span></span>
<span><span class="co">#&gt; [37] vctrs_0.6.5       evaluate_1.0.3    glue_1.8.0        farver_2.1.2     </span></span>
<span><span class="co">#&gt; [41] ragg_1.3.3        colorspace_2.1-1  rmarkdown_2.29    tools_4.4.2      </span></span>
<span><span class="co">#&gt; [45] pkgconfig_2.0.3   htmltools_0.5.8.1</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jessie Maier, Manuel Kleiner.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
