#' Plot read coverage graphs of contigs with predicted active prophages in the whole-community read coverages
#'
#' Plot the read coverages of a contig and its associated shape-match for each contig with a predicted active prophage.
#'
#' @param pileup A table containing contig names, coverages averaged over 100bp windows, and contig positions associated with mapping whole-community reads to whole-community contigs
#' @param ProActive_shapelist A list containing shape information associated with all contigs containing a potential active prophage. The second, third or fourth item in the list output from ProActive()
#' @param ProActiveResults The result from ProActive()
#' @param elevation_filter Default is NA. The user may define an elevation factor in which any plots with elevated read coverages below the elevation factor are not displayed
#' @param cleanup If pileup files are generated by BBMap's pileup.sh, the files output from pileup.sh can be input directly into ProActive with cleanup=TRUE. If your pileup files are NOT generated with BBMap's pileup.sh, set cleanup=FALSE. The user is responsible for data cleaning/formatting if this option is used.
#'
#' @export
#'
#' @examples
#' \dontrun{
#' ProActivePredictionPlots(whole_commreadcovs, ProActiveResults, ProActiveResults$ExtraConfidentPredictions,cleanup=TRUE)
#' }
ProActivePredictionPlots <- function(pileup, ProActiveResults, ProActive_shapelist, elevation_filter=NA,cleanup=TRUE) {
  position <- coverage <- NULL
  if (length(ProActiveResults)==9){
    windowsize <- ProActiveResults[[7]]
    mode <- ProActiveResults[[8]]
    chunksize <- ProActiveResults[[9]]
    ProActive_summarydf <- ProActiveResults[[1]]
  }else {
    windowsize <- ProActiveResults[[6]]
    mode <- ProActiveResults[[7]]
    chunksize <- ProActiveResults[[8]]
    ProActive_summarydf <- ProActiveResults[[1]]
  }
  if(cleanup==TRUE){
    pileup <- readcovdf_formatter(pileup, mode)
  }
  plots <- list()
  ref_names <- c()
  if (mode=="genome"){
    pileup <- GenomeChunks(pileup, chunksize)
  }
  for (i in seq(1,length(ProActive_shapelist),1)) {
    ref_name <- ProActive_shapelist[[i]][[8]]
    ref_names <- c(ref_names, ref_name)
    match_info <- ProActive_summarydf[which(ProActive_summarydf[,1]==ref_name),]
    prediction <- match_info[,2]
    elev_ratio <- ifelse(prediction=="None","NA",round(match_info[,4], digits=3))
    if (is.na(elevation_filter) == FALSE) {
      if (elev_ratio < elevation_filter) next
    }
    microbial_subset <- pileup[which(pileup[,1] == ref_name),]
    microbial_subset <- windowsize_func(microbial_subset,windowsize, mode)
    shape <- shape_builder_func(microbial_subset, ProActive_shapelist, i)
    shape_match <- cbind(microbial_subset, shape)
    match_length <- match_info[,7]
    plot <- ggplot(data=shape_match, aes(x=position, y=coverage))+
            geom_area(data=shape_match, aes(x=position, y=coverage), fill="deepskyblue3") +
            geom_line(y=shape, linewidth=1)+
            labs(title=ref_name,subtitle=paste("Matching-region size (bp):", match_length, "Elevation ratio:", elev_ratio), x="contig position") +
            theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                  panel.background = element_blank(), axis.line = element_line(colour = "black"),text = element_text(size = 15))
    plots[[i]] <- plot
  }
  names(plots) <- ref_names
  return(plots)
}
